app/src/store/__tests__/audioDevices.test.ts

```diff
@@ -44,73 +56,352 @@ jest.mock('@twilio/voice-react-native-sdk', () => {
+      matchDispatchedActions(dispatchedActions, [
+        audioDevices.updateAudioDevices.pending,
+        audioDevices.updateAudioDevices.fulfilled,
+      ]);
+    });
+
+    it('handles no selected device', async () => {
+      const mockAudioDevices = [
+        { uuid: '1111', type: 'speaker', name: 'device1' },
+        { uuid: '2222', type: 'bluetooth', name: 'device2' },
+        { uuid: '3333', type: 'earpiece', name: 'device3' },
+      ];
+
+      await store.dispatch(
+        audioDevices.updateAudioDevices({
+          audioDevices: mockAudioDevices as any,
+          selectedDevice: undefined,
+        }),
+      );
+
+      expect(store.getState().voice.audioDevices).toEqual({
+        status: 'fulfilled',
+        audioDevices: mockAudioDevices,
+        selectedDevice: undefined,
+      });
+      matchDispatchedActions(dispatchedActions, [
+        audioDevices.updateAudioDevices.pending,
+        audioDevices.updateAudioDevices.fulfilled,
+      ]);
+    });
+
+    it('handles audio device undefined', async () => {
+      await store.dispatch(
+        audioDevices.updateAudioDevices({ audioDevices: undefined as any }),
+      );
+
+      expect(store.getState().voice.audioDevices).toEqual({
+        status: 'rejected',
+        reason: 'AUDIO_DEVICES_UNDEFINED',
+      });
+      matchDispatchedActions(dispatchedActions, [
+        audioDevices.updateAudioDevices.pending,
+        audioDevices.updateAudioDevices.rejected,
+      ]);
+    });
+
+    it('handles uuid undefined', async () => {
+      const mockAudioDevices = [
+        { uuid: '1111', type: 'speaker', name: 'device1' },
+        { uuid: undefined, type: 'bluetooth', name: 'device2' },
+        { uuid: '3333', type: 'earpiece', name: 'device3' },
+      ];
+
+      await store.dispatch(
+        audioDevices.updateAudioDevices({
+          audioDevices: mockAudioDevices as any,
+        }),
+      );
+
+      expect(store.getState().voice.audioDevices).toEqual({
+        status: 'rejected',
+        reason: 'AUDIO_DEVICE_UUID_UNDEFINED',
+      });
+      matchDispatchedActions(dispatchedActions, [
+        audioDevices.updateAudioDevices.pending,
+        audioDevices.updateAudioDevices.rejected,
+      ]);
+    });
+
+    it('handles an unexpected error', async () => {
+      const mockAudioDevices = [
+        { uuid: '1111', type: 'speaker', name: 'device1' },
+        { uuid: '2222', type: 'bluetooth', name: 'device2' },
+        { uuid: '3333', type: 'earpiece', name: 'device3' },
+      ];
+
+      const error = new Error('bye');
+      delete error.stack;
+      jest.spyOn(audioDeviceMap, 'set').mockImplementationOnce(() => {
+        throw error;
+      });
+
+      const result = await store.dispatch(
+        audioDevices.updateAudioDevices({
+          audioDevices: mockAudioDevices as any,
+        }),
+      );
+      if (!audioDevices.updateAudioDevices.rejected.match(result)) {
+        throw new Error('dispatch result should be rejected');
```

Comment:
why `throw new Error` instead of use an `expect`?

Reply:
Using an expect would look something like this
```
expect(audioDevices.updateAudioDevices.rejected.match(result)).toBe(true);
```
If this expectation fails, then the failure message in jest would just be something simple like
```
expected true, got false
```
or something, which isn't the most helpful error message in the world. If we throw an error with a specific error message, then I think it helps debugging unit tests.